libyul/optimiser/NameDispenser.cpp

```diff
@@ -42,6 +42,21 @@ YulString NameDispenser::newName(YulString _nameHint, YulString _context)
 	// Shortening rules: Use a suffix of _prefix and a prefix of _context.
 	YulString prefix = _nameHint;
 
+	// If the name itself has a suffix supplied by user, get rid of it
+	// so that we have no conflicts and the order of suffix is in order.
+	std::string nameHintStr = _nameHint.str();
+	size_t underscore = nameHintStr.find_last_of("_");
+	if (underscore != std::string::npos && underscore != 0)
+	{
+		try
+		{
+			std::stoi(nameHintStr.substr(underscore + 1));
+			prefix = YulString(nameHintStr.substr(0, underscore));
+		} catch (std::invalid_argument &e) {
```

Comment:
Maybe we should avoid throwing and catching exceptions here by checking whether everything after the underscore is a digit?